trie/triedb/pathdb/database.go

```diff
@@ -0,0 +1,396 @@
+// to only keep 128 diff layers in memory by default.
+//
+// The passed in maps(nodes, states) will be retained to avoid copying everything.
+// Therefore, these maps must not be changed afterwards.
+func (db *Database) Update(root common.Hash, parentRoot common.Hash, block uint64, nodes *trienode.MergedNodeSet, states *triestate.Set) error {
+	// Hold the lock to prevent concurrent mutations.
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	// Short circuit if the database is in read only mode.
+	if db.readOnly {
+		return errSnapshotReadOnly
+	}
+	if err := db.tree.add(root, parentRoot, block, nodes, states); err != nil {
+		return err
+	}
+	// Keep 128 diff layers in the memory, persistent layer is 129th.
+	// - head layer is paired with HEAD state
+	// - head-1 layer is paired with HEAD-1 state
+	// - head-127 layer(bottom-most diff layer) is paired with HEAD-127 state
+	// - head-128 layer(disk layer) is paired with HEAD-128 state
+	return db.tree.cap(root, maxDiffLayers)
+}
+
+// Commit traverses downwards the layer tree from a specified layer with the
+// provided state root and all the layers below are flattened downwards. It
+// can be used alone and mostly for test purposes.
+func (db *Database) Commit(root common.Hash, report bool) error {
+	// Hold the lock to prevent concurrent mutations.
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	// Short circuit if the database is in read only mode.
+	if db.readOnly {
+		return errSnapshotReadOnly
+	}
+	return db.tree.cap(root, 0)
+}
+
+// Reset rebuilds the database with the specified state as the base.
+//
+//   - if target state is empty, clear the stored state and all layers on top
+//   - if target state is non-empty, ensure the stored state matches with it
+//     and clear all other layers on top.
+func (db *Database) Reset(root common.Hash) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	// Short circuit if the database is in read only mode.
+	if db.readOnly {
+		return errSnapshotReadOnly
+	}
+	batch := db.diskdb.NewBatch()
+	root = types.TrieRootHash(root)
+	if root == types.EmptyRootHash {
+		// Empty state is requested as the target, nuke out
+		// the root node and leave all others as dangling.
+		rawdb.DeleteAccountTrieNode(batch, nil)
+	} else {
+		// Ensure the requested state is existent before any
+		// action is applied.
+		_, hash := rawdb.ReadAccountTrieNode(db.diskdb, nil)
+		if hash != root {
+			return fmt.Errorf("state is mismatched, local: %x, target: %x", hash, root)
+		}
+	}
+	// Mark the disk layer as stale before applying any mutation.
+	db.tree.bottom().markStale()
+
+	// Drop the stale state journal in persistent database and
+	// reset the persistent state id back to zero.
+	rawdb.DeleteTrieJournal(batch)
+	rawdb.WritePersistentStateID(batch, 0)
+	if err := batch.Write(); err != nil {
+		return err
+	}
+	// Clean up all state histories in freezer. Theoretically
+	// all root->id mappings should be removed as well. Since
```

Comment:
If we don't remove the root->id mappings, won't that potentially cause `Recoverable` to return true for non existing  states?

Reply:
Ah, I guess the check against the bottom layer ensure that only "overwritten" states can be recovered, but dangling junk cannot. Is my reasoning correct here?

Reply:
> If we don't remove the root->id mappings, won't that potentially cause Recoverable to return true for non existing states?

Nope, Recoverable check contains two steps:
- Ensure the target is known by checking the relevant state id
- Ensure all the histories from id+1 until disk layer are all present

If we leave root->id mappings in disk, the non-existing state will be "known", but it lacks of corresponding state history, so it's still Unrecoverable.